### PR TITLE
Fix: REST API "download=False" error 

### PR DESCRIPTION
**Pull request by eimrek**
My proposal for #5574

for context, see the discussion in the issue.

### CODE REVIEW

File: aiida/restapi/resources.py [hunk 1]
```diff
@@ -458,7 +458,7 @@ def get(self, id=None, page=None):  # pylint: disable=redefined-builtin,invalid-
 
                 if query_type == 'download' and download not in ['false', 'False', False] and results:
                     if results['download']['status'] == 200:
-                        data = results['download']['data']
+                        data = results['download']['data'].encode('utf-8')
```
Comment (ltalirz) [inline aiida/restapi/resources.py:461]: Can you explain why this change is necessary?

Didn't this work before?
Is `node._exportcontent` involved here?

Comment (eimrek) [reply aiida/restapi/resources.py:461]: Thanks. For context, check "Additional context" on https://github.com/aiidateam/aiida-core/issues/5574.

I'm not sure how it worked before, as `node._exportcontent` says it always returns a binary string.

I'll do some debugging

**Comment by eimrek** [comment]
Ok, I figured out the source of this problem:

The `JSONEncoder` imported here

https://github.com/aiidateam/aiida-core/blob/cb82327e5e4587b8752674a6b57e64613a5e11f2/aiida/restapi/common/utils.py#L15

in flask 1.x was based on `simplejson`, that inherently converts bytes objects into strings. In flask 2.0.0, it was redesigned based on the default python `json`. See here: https://flask.palletsprojects.com/en/2.0.x/changes/ and here https://github.com/pallets/flask/issues/3555

As a result, in Aiida 2.0 REST (where flask 1 was updated to flask 2), bytes objects are not supported any more.

I implemented a conversion of a bytes array into `str` in the `CustomJSONEncoder`. In principle there are other way of doing this as well (e.g. by checking `isinstance(o, bytes)`). I can change it, if some other implementation is preferable.

Similar issue is probably also the source of #5575 but this PR doesn't fix that. i will investigate that in a separate PR.

I'll try to add a test as well.

**Review by giovannipizzi — APPROVED** [review]
Thanks!

**Review by unkcpz — CHANGES_REQUESTED** [review]
I think it makes sense to combine https://github.com/aiidateam/aiida-core/pull/5573 to this PR.

**Comment by eimrek** [comment]
Thanks @unkcpz. I disagree, as the source issues are different. #5573 fixes a bug that has been there forever, and affects how boolean query parameters are parsed. The issue of this PR is in the encoding of the binary string to JSON, that was introduced by the aiida 2.0 upgrade. Although the issue comes out in case of `download=False`, that might seem similar, it probably also affects other parts of the restapi.

I think it makes sense to isolate the code changes that fix specific issues, instead of merging everything together. the current PR fixes the linked issue (and the PR of #5573 does not affect this).

 PR #5573 fixes a separate issue, namely, that lowercase "false" in the query parameters is interpreted as True, but could be bypassed by specifying an uppercase "False" instead. I suspect this is the reason is it was not fixed for over 3 years.

File: aiida/restapi/common/utils.py [hunk 2]
```diff
@@ -56,6 +56,12 @@ def default(self, o):
                                    o.hour).zfill(2), str(o.minute).zfill(2),
                                          str(o.second).zfill(2)])
 
+        # To support bytes objects, try to decode to a string
+        try:
+            return o.decode('utf-8')
+        except (UnicodeDecodeError, AttributeError):
+            pass
```
Comment (unkcpz) [inline aiida/restapi/common/utils.py:63]: I think an `elif` for bytes type makes more sense.

**Review by unkcpz — CHANGES_REQUESTED** [review]
@eimrek sorry for the back and forth. One more minor request, others are all good. Could you then rebase and make a detailed commit message? Thanks.

File: aiida/restapi/common/utils.py (hunk 2 above)
Comment (eimrek) [reply aiida/restapi/common/utils.py:63]: This discussion here seems to prefer the duck-typing version: https://stackoverflow.com/questions/34869889/what-is-the-proper-way-to-determine-if-an-object-is-a-bytes-like-object-in-pytho

**Review by unkcpz — APPROVED** [review]
